optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,371 @@
+                .penalize("Period penalty after", HardSoftScore.ONE_HARD,
+                          (periodPenalty, leftExam, rightExam) ->
+                                  leftExam.getTopic().getStudentSize() + rightExam.getTopic().getStudentSize());
+    }
+
+    //    rule "roomPenaltyExclusive"
+//    when
+//    $roomPenalty : RoomPenalty(
+//            roomPenaltyType == RoomPenaltyType.ROOM_EXCLUSIVE,
+//            $topic : topic
+//    )
+//    $leftExam : Exam(topic == $topic, $room : room, room != null, $period : period, period != null)
+//    $rightExam : Exam(room == $room, period == $period, topic != $topic, $otherTopic : topic)
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext,
+//            - ($topic.getStudentSize() + $otherTopic.getStudentSize()));
+//    end
+    private Constraint roomPenaltyExclusive(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RoomPenalty.class)
+                .filter(roomPenalty -> roomPenalty.getRoomPenaltyType() == RoomPenaltyType.ROOM_EXCLUSIVE)
+                .join(Exam.class,
+                      equal(RoomPenalty::getTopic, Exam::getTopic))
+                .filter((roomPenalty, leftExam) -> leftExam.getPeriod() != null && leftExam.getRoom() != null)
+                .join(Exam.class,
+                      equal((roomPenalty, leftExam) -> leftExam.getRoom(), Exam::getRoom),
+                      equal((roomPenalty, leftExam) -> leftExam.getPeriod(), Exam::getPeriod))
+                .filter((roomPenalty, leftExam, rightExam) -> leftExam.getTopic().equals(rightExam.getTopic()))
+                .penalize("Room penalty exclusive", HardSoftScore.ONE_HARD,
+                          (periodPenalty, leftExam, rightExam) -> leftExam.getTopic().getStudentSize() + rightExam.getTopic().getStudentSize());
+    }
+
+    //    // Two exams in a row which share students
+//    rule "twoExamsInARow"
+//    when
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $leftDayIndex : dayIndex, $leftPeriodIndex : periodIndex, period != null)
+//    $rightExam : Exam(topic == $rightTopic, dayIndex == $leftDayIndex,
+//                      Math.abs($leftPeriodIndex - periodIndex) == 1)
+//    then
+//        scoreHolder.penalize(kcontext, $topicConflict.getStudentSize());
+//    end
+    private Constraint twoExamsInARow(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
```

Comment:
Perhaps could be replaced by `ifExists()`, but then the constraint may no longer be functionally equivalent to the DRL. `FULL_ASSERT` would tell.